subprojects/core/src/main/java/org/gradle/internal/vfs/DefaultWatchingVirtualFileSystem.java

```diff
@@ -0,0 +1,163 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.vfs;
+
+import com.google.common.collect.Interner;
+import com.sun.nio.file.SensitivityWatchEventModifier;
+import org.gradle.api.internal.changedetection.state.WellKnownFileLocations;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.hash.FileHasher;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.vfs.impl.DefaultVirtualFileSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+public class DefaultWatchingVirtualFileSystem extends DefaultVirtualFileSystem implements WatchingVirtualFileSystem, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWatchingVirtualFileSystem.class);
+
+    private final WellKnownFileLocations wellKnownFileLocations;
+    private WatchService watchService;
+
+    public DefaultWatchingVirtualFileSystem(
+        FileHasher hasher,
+        Interner<String> stringInterner,
+        Stat stat,
+        WellKnownFileLocations wellKnownFileLocations,
+
+        CaseSensitivity caseSensitivity,
+        String... defaultExcludes
+    ) {
+        super(hasher, stringInterner, stat, caseSensitivity, defaultExcludes);
+        this.wellKnownFileLocations = wellKnownFileLocations;
+    }
+
+    @Override
+    public void startWatching() {
+        if (watchService != null) {
+            throw new IllegalStateException("Watch service already started");
+        }
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+        Set<File> visited = new HashSet<>();
+        root.get().visitCompleteSnapshots(snapshot -> {
+            String absolutePath = snapshot.getAbsolutePath();
+            File file = new File(absolutePath);
+            if (wellKnownFileLocations.isImmutable(absolutePath)) {
+                return;
+            }
+            try {
+                if (snapshot.getType() == FileType.Directory) {
```

Comment:
Can you add some comment here explaining why we watch the direct parent of any complete snapshot? Some explanation how the watching is expected to work would be great.